src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs

```diff
@@ -90,12 +90,14 @@ public DurableTaskExtension()
         /// <param name="nameResolver">The name resolver to use for looking up application settings.</param>
         /// <param name="orchestrationServiceFactory">The factory used to create orchestration service based on the configured storage provider.</param>
         /// <param name="durableHttpMessageHandlerFactory">The HTTP message handler that handles HTTP requests and HTTP responses.</param>
+        /// <param name="lifeCycleNotificationHelper">The lifecycle notification helper used for custom orchestration tracking.</param>
         public DurableTaskExtension(
             IOptions<DurableTaskOptions> options,
             ILoggerFactory loggerFactory,
             INameResolver nameResolver,
             IOrchestrationServiceFactory orchestrationServiceFactory,
-            IDurableHttpMessageHandlerFactory durableHttpMessageHandlerFactory = null)
+            IDurableHttpMessageHandlerFactory durableHttpMessageHandlerFactory = null,
+            ILifeCycleNotificationHelper lifeCycleNotificationHelper = null)

```

Comment:
Dependency Injection supports constructor default value, so if you don't add a singleton service, the default value of null is passed.

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-2.1#constructor-injection-behavior

Reply:
I see now why you are handling this with `null` instead of registering the service in `DurableTaskJobHostConfigurationExtensions.AddDurableTask()`. 

It is because the version of `ILifeCycleNotificationHelper` that we use is dependent on the config in `IOptions<DurableTaskOptions>()`. Normally, the approach we would use with this is a factory interface, but considering we already have a method that initializes the correct `ILifeCycleNotificationHelper`, this seems simpler. 

It also allows the customer to directly inject an `ILifeCycleNotificationHelper` without having to create an `ILifeCycleNotificationHelperFactory`. 

Because of this, I would say I prefer your implementation.